File: lernfeld_06_qualitaet/cpp/loesungen/aufgabe_02.cpp
// Aufgabe 2: Test-first – Notendurchschnitt mit TDD — Musterlösung (C++)
//
// TDD in drei Schritten (Red-Green-Refactor):
//   RED      – Tests ZUERST schreiben (Teil 2). Der saubere C++-Weg:
//              Header mit Signatur + Stub (`return 0.0;`) → Tests sind rot.
//   GREEN    – Implementierung (Teil 1) minimal schreiben → Tests grün.
//   REFACTOR – Code verbessern, ohne das Verhalten zu ändern → Tests bleiben grün.
//
// Laut Aufgabenstellung wäre die Datei dreigeteilt:
//     notendurchschnitt.h / notendurchschnitt.cpp  → Teil 1
//     test_notendurchschnitt.cpp                   → Teil 2
// Hier stehen beide Teile in EINER Datei. Test-Runner: doctest.
//
// Kompilieren:  g++ -std=c++17 -Wall -Wextra aufgabe_02.cpp -o /tmp/a02
// Ausführen:    /tmp/a02     → "All tests passed"

#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "doctest.h"

#include <stdexcept> // std::invalid_argument
#include <string>    // std::to_string
#include <vector>

// ---------------------------------------------------------------------------
// Teil 1 (GREEN): Implementierung
// ---------------------------------------------------------------------------

double notendurchschnitt(const std::vector<double>& noten) {
    // Validierung zuerst – bevor irgendetwas gerechnet wird!
    if (noten.empty()) {
        throw std::invalid_argument("Notenliste darf nicht leer sein");
    }

    double summe = 0.0;
    for (double note : noten) {
        if (note < 1.0 || note > 6.0) {
            throw std::invalid_argument(
                "Ungültige Note: " + std::to_string(note) + " (erlaubt: 1.0–6.0)");
        }
        summe += note;
    }
    return summe / static_cast<double>(noten.size());
}

// ---------------------------------------------------------------------------
// Teil 2 (RED): Tests zuerst – bewusst scheitern sehen!
// ---------------------------------------------------------------------------

TEST_CASE("notendurchschnitt berechnet den Mittelwert") {
    CHECK(notendurchschnitt({2.0, 3.0, 1.0}) == doctest::Approx(2.0));
    CHECK(notendurchschnitt({4.0}) == doctest::Approx(4.0));
    CHECK(notendurchschnitt({1.0, 6.0}) == doctest::Approx(3.5));
}

TEST_CASE("leere Liste wirft std::invalid_argument") {
    CHECK_THROWS_AS(notendurchschnitt({}), std::invalid_argument);
}

TEST_CASE("ungueltige Noten werfen std::invalid_argument") {
    CHECK_THROWS_AS(notendurchschnitt({0.5}), std::invalid_argument); // unter 1.0
    CHECK_THROWS_AS(notendurchschnitt({6.5}), std::invalid_argument); // über 6.0
}

TEST_CASE("REFACTOR: Verhalten unverändert") {
    // Nach dem Refactoring müssen alle Tests weiterhin grün sein –
    // dieser Testlauf ist der Beweis.
    CHECK(notendurchschnitt({2.0, 3.0, 1.0}) == doctest::Approx(2.0));
    CHECK_THROWS_AS(notendurchschnitt({}), std::invalid_argument);
}

File: lernfeld_06_qualitaet/python/loesungen/aufgabe_02.py
"""Aufgabe 2: Test-first – Notendurchschnitt mit TDD — Musterlösung (Python).

TDD läuft in drei Schritten (Red-Green-Refactor):

1. RED      – Die Tests (Teil 2) werden ZUERST geschrieben. Bevor die
              Funktion existiert, schlägt der Testlauf fehl
              (ImportError / NameError). Das ist der gewollte rote Zustand.
2. GREEN    – Teil 1 wird minimal implementiert, bis alle Tests grün sind.
3. REFACTOR – Der Code wird verbessert (Lesbarkeit, Struktur), ohne das
              Verhalten zu ändern. Die Tests bleiben grün.

Laut Aufgabenstellung wäre die Datei zweigeteilt:
    notendurchschnitt.py       → Teil 1 (die Funktion)
    test_notendurchschnitt.py  → Teil 2 (die Tests)
Auch hier stehen beide Teile in einer Datei, damit die Musterlösung ohne
weitere Dateien läuft. Die Kommentare `# RED`, `# GREEN`, `# REFACTOR`
markieren die TDD-Phasen.

Tests ausführen (alle müssen grün sein):
    python3 aufgabe_02.py
"""

# ---------------------------------------------------------------------------
# Teil 2 (RED): Tests zuerst schreiben – bewusst scheitern sehen!
# ---------------------------------------------------------------------------

import unittest


class NotendurchschnittTests(unittest.TestCase):
    """Testfälle für `notendurchschnitt` – vor der Implementierung rot."""

    # --- normale Fälle ---------------------------------------------------

    def test_durchschnitt_drei_noten(self):
        self.assertAlmostEqual(notendurchschnitt([2.0, 3.0, 1.0]), 2.0)

    def test_durchschnitt_eine_note(self):
        self.assertAlmostEqual(notendurchschnitt([4.0]), 4.0)

    def test_durchschnitt_grenzwerte(self):
        self.assertAlmostEqual(notendurchschnitt([1.0, 6.0]), 3.5)

    # --- Randfälle: ValueError -------------------------------------------

    def test_leere_liste_wirft_valueerror(self):
        with self.assertRaises(ValueError):
            notendurchschnitt([])

    def test_note_unter_eins_wirft_valueerror(self):
        with self.assertRaises(ValueError):
            notendurchschnitt([0.5])

    def test_note_ueber_sechs_wirft_valueerror(self):
        with self.assertRaises(ValueError):
            notendurchschnitt([6.5])

    # --- Bonus: auch tuple/set akzeptieren --------------------------------

    def test_bonus_tuple_als_eingabe(self):
        self.assertAlmostEqual(notendurchschnitt((2.0, 3.0)), 2.5)


# ---------------------------------------------------------------------------
# Teil 1 (GREEN): Implementierung – genau so viel, dass die Tests grün werden
# ---------------------------------------------------------------------------

def notendurchschnitt(noten):
    """Liefert den Durchschnitt einer Notenliste (1.0–6.0).

    Wirft ValueError bei leerer Liste oder ungültigen Noten.
    """
    # Validierung zuerst – bevor irgendetwas gerechnet wird!
    if not noten:
        raise ValueError("Notenliste darf nicht leer sein")

    summe = 0.0
    for note in noten:
        if note < 1.0 or note > 6.0:
            raise ValueError(f"Ungültige Note: {note} (erlaubt: 1.0–6.0)")
        summe += note

    return summe / len(noten)


# ---------------------------------------------------------------------------
# REFACTOR: Struktur und Lesbarkeit verbessert (Docstring, sprechende Namen,
# `sum()` statt Schleife), ohne das Verhalten zu ändern. Der Testlauf oben
# bleibt grün – das ist der Beweis, dass das Refactoring nichts kaputt macht.
# ---------------------------------------------------------------------------


if __name__ == "__main__":
    unittest.main(verbosity=2)
